app/code/community/Bolt/Boltpay/controllers/ApiController.php

```diff
@@ -237,4 +237,61 @@ protected function getCaptureAmount($transaction) {
 
         return null;
     }
+
+    /**
+     * Handles failed payment web hooks.  It attempts to cancel a specified pre-auth order
+     * in addition to invalidating the cache associated with that orders session.
+     *
+     * @param int   $parentQuoteId  the ID of the session quote whose order should be cancelled.
+     *
+     * @throws Zend_Controller_Response_Exception if there is an unexpected error in sending a response
+     */
+    private function handleFailedPaymentHook($parentQuoteId) {
+        /** @var Bolt_Boltpay_Model_Order $orderModel */
+        $orderModel = Mage::getModel('boltpay/order');
+        $order =  $orderModel->getOrderByParentQuoteId($parentQuoteId);
+        if (!$order->isObjectNew()) {
+            $orderModel->removePreAuthOrder($order);
+        }
+        ////////////////////////////////////////////////////////////////////
+        /// We treat Bolt initiated failed payment cancels to be the same as a directive
+        /// to expire the cached immutable quote that is stored in the session,
+        /// otherwise, the Bolt checkout could result in a locked state where the end
+        /// user will repeatedly be told that his cart has expired and to refresh. However,
```

Comment:
just to clarify I believe `failed` hook is only called for failed payment. If preauth endpoint returns 'cart expired' error, Bolt will not call hook

Reply:
Yes. I know.  There is no need to call a hook at that point as the order will not have been created and there is nothing to cancel.  We are talking here after the Bolt server has received the create_order success reply with the success URL